### ISSUE TITLE
get() and param() functions stall on RFC 6570 templated query components

### ISSUE THREAD

**Issue by Jitsusama** [rest-assured/rest-assured#954]
RFC 6570 defines query components in section [2.2](https://tools.ietf.org/html/rfc6570#section-2.2). These allow an HTTP URI template to show query parameters that are possible for an endpoint with the name of these parameters specified.

When I try to use such templates generated by Spring's HATEOAS library (which specifically supports RFC 6570) in either a `given().when().get()` or `given().param()` call running within Spring's test runner, the call stalls indefinitely. I've tried the same without Spring involved at all and experienced the same issue.

Here's a very simple example that I used to reproduce the issue:

```java
package blah;

import org.junit.Test;
import static io.restassured.RestAssured.given;

class Rfc6570QueryComponentTemplateTest {
    @Test public void simpleTest() {
        given().
            when().get("http://localhost/things{?olderThan}", "3days").
            then().log();
    }
}
```

Running this example through JUnit 4 stalls indefinitely. I've tested this with version 3.0.5 of the rest-assured library obtained through Maven.

In summary; the issue is definitely there, supporting such templates would be a nice feature to adopt as a number of HATEOAS APIs exist out there that take advantage of RFC 6570 URI templates, and even if there is no desire to implement this functionality, having the test stall indefinitely is probably not the desired result for an unsupported operation.

Please let me know if there is any further information that I should provide and thanks so much for this great library!

**Comment by Jitsusama** [comment]
I probably should mention that the expected behaviour is when `get()` is called with a templated URI and value, that it would expand it in a way comformant to RFC 6750. Here's an example:

For templated string `http://localhost/things{?olderThan}`, when passing in `3days` as a value, should be expanded into `http://localhost/things?olderThan=3days`.

**Comment by johanhaleby** [comment]
Templates in REST Assured are written like this:

```java
@Test public void simpleTest() {
        given().
            when().get("http://localhost/things?olderThan={amountAndUnit}", "3days").
            then().log();
    }
```

I've chosen this because it is (was?) the way that Spring does and it seems to be familiar with many users.

**Comment by johanhaleby** [comment]
I have no intention of adding support for RFC6570 although I wouldn't mind if it were supported :). If you're willing to help out and add support for it that would be nice (given that backward-compatibility) is not broken.

**Comment by Jitsusama** [comment]
I don't mind the idea of introducing it. Do you have any overall design documentation as to how your code's logic flows, or any pointers at what area of code I can look at for finding where this logic could be introduced? I've already tried poking, but I got lost pretty quick.

I should mention that Spring HATEOAS actually generates templates exactly the same as I posted, which is how I came to butting my head against this. It seems that the Spring group has 2 separate ideas, which HATEOAS being the one that decided to support the RFC 6570 standardized approach (which I like because of how flexible it is).

**Comment by johanhaleby** [comment]
Sorry there's no design document, you'd have to look in the code (and I know that this can be hard so sorry about that). The tempting stuff is done manually (i.e. I've coded it myself) now, perhaps there's some library that implements both that we could use or take inspiration from instead?

**Comment by Jitsusama** [comment]
https://github.com/damnhandy/Handy-URI-Templates seems to be a good library. I've played with it a bit already. It handles the creation and decomposition of templates and fully implement RFC 6570.

**Comment by johanhaleby** [comment]
Ok cool. But now I begin to wonder if this is really worth it. Perhaps it's better to simply use that library if you want to have RFC6570 support instead of baking it in the REST Assured? You could just pass the uri generated by the library to REST Assured.

**Comment by Jitsusama** [comment]
Honestly, that's what I'm doing now. I'm more thinking of making this library simpler to use for other people trying to handle URI templates (and to make it simpler for myself). Since the library has quasi-support right now, it's rather counter-intuitive to have some templates work and not others.

**Comment by Jitsusama** [comment]
Also, compare these use examples;

* Using Handy-URI-Templates:

    ```java
    package blah;

    import com.damnhandy.uri.template.UriTemplate;
    import io.restassured.RestAssured;
    import org.junit.Test;

    class TemplateTests {
        @Test public void simpleTest() {
            String requestUri = UriTemplate
                .fromTemplate("http://localhost/things{?olderThan}")
                .set("olderThan", "3days")
                .expand();
            given().
                when().get(requestUri).
                then().log();
        }
    }
    ```

* Having it native

    ```java
    package blah;

    import io.restassured.RestAssured;
    import org.junit.Test;

    class TemplateTests {
        @Test public void simpleTest() {
            given().
                when().get("http://localhost/things{?olderThan)", "3days").
                then().log();
        }
    }
    ```

I really like the native solution's look much better.

**Comment by johanhaleby** [comment]
It does indeed look better but the question is whether it's worth the extra complexity added to the project and the extra dependency. I'm sure that there are other libraries for URI templating as well out there that I assume some people would like to use, it wouldn't make sense to integrate all of them. Perhaps there can also be conflicts?

REST Assured is 7 years old and predates the RFC and you're the first person that have requested support for this afaik. Thus I suspect that most people are fine with the templating that's used now.

WDYT?

**Comment by Jitsusama** [comment]
You're probably right. Just because I would like to have it does not mean that others necessarily would feel the same. I have found at least 3 Java libraries that have various levels of support for RFC 6570.

I would argue that it looks like HATEOAS styled REST APIs are on the rise, and a big part of these APIs is being able to discover links as well as link parameters via RESTful queries. I would expect the demand for this to rise, but then again, maybe other people are already realizing this and are resorting to other template libraries as is and not feeding similar comments back into this project?

**Comment by johanhaleby** [comment]
> I would argue that it looks like HATEOAS styled REST APIs are on the rise, and a big part of these APIs is being able to discover links as well as link parameters via RESTful queries

That would be good if true :) 

> I would expect the demand for this to rise, but then again, maybe other people are already realizing this and are resorting to other template libraries as is and not feeding similar comments back into this project?

If we notice that the demand for this increase (and especially if more people request it) then we can revisit our decision? For example it might be possible to solely RFC 6570 templating in version 4.0 if we see that it's getting popular.

**Comment by Jitsusama** [comment]
That sounds like a good plan.

**Comment by johanhaleby** [comment]
Ok, I'm closing it for now
